impl/src/testing.rs

```diff
@@ -56,3 +96,107 @@ pub fn make_workspace(
+pub fn mock_remote_crate(name: &str, version: &str, mock_server: &MockServer) -> TempDir {
+  // Crate info mock response
+  mock_server.mock(|when, then| {
+    when.method(GET).path(format!("/api/v1/crates/{}", name));
+    then.status(200).json_body(json!({
+        "crate": {
+            "id": name,
+            "name": name,
+            "versions": [
+                283716
+            ],
+        },
+        "versions": [
+            {
+                "id": 283716,
+                "crate": name,
+                "num": version,
+                "dl_path": format!("/api/v1/crates/{}/{}/download", name, version),
+            }
+        ],
+    }));
+  });
+
+  // Create archive contents
+  let dir = tempfile::TempDir::new().unwrap();
+  let tar_path = dir.as_ref().join(format!("{}.tar.gz", name));
+  {
+    std::fs::create_dir_all(dir.as_ref().join("archive")).unwrap();
+    std::fs::File::create(dir.as_ref().join("archive/test")).unwrap();
+    std::fs::write(dir.as_ref().join("archive/Cargo.toml"), named_toml(name, version)).unwrap();
+    std::fs::write(dir.as_ref().join("archive/Cargo.lock"), named_lock(name, version)).unwrap();
+
+    let tar_gz: std::fs::File = std::fs::File::create(&tar_path).unwrap();
+    let enc = flate2::write::GzEncoder::new(tar_gz, Compression::default());
+    let mut tar = tar::Builder::new(enc);
+    tar
+      .append_dir_all(
+        format!("{}-{}", name, version),
+        dir.as_ref().join("archive"),
+      )
+      .unwrap();
+  }
+
+  // Create download mock response
+  mock_server.mock(|when, then| {
+    when
+      .method(GET)
+      .path(format!("/api/v1/crates/{}/{}/download", name, version));
+    then
+      .status(200)
+      .header("content-type", "application/x-tar")
+      .body_from_file(&tar_path.display().to_string());
+  });
+
+  dir
+}
+
+/** A helper macro for passing a `crates` to  `mock_crate_index` */
+#[macro_export]
+macro_rules! index_crates {
+  ($pkgs:expr) => {{
+    $pkgs
+      .iter()
+      .map(|(k, v)| (k.to_string(), v.to_string()))
+      .collect()
+  }};
+}
+
+/** Create a mock cache in a temporary direcotry that contains a set of given crates */
+pub fn mock_crate_index(crates: &HashMap<String, String>) -> TempDir {
+  let index_url = TempDir::new().unwrap();
+
+  for (name, version) in crates {
+    assert!(
+      name.len() >= 5,
+      "Mock crate names should be at least 5 characters"
+    );
+    let path = index_url
+      .as_ref()
+      .join(&name.as_str()[0..2])
+      .join(&name.as_str()[2..4])
+      .join(name);
+
+    std::fs::create_dir_all(path.parent().unwrap()).unwrap();
+    std::fs::write(
+      path,
+      json!({
+        "name": name,
+        "vers": version,
+        "deps": [],
+        "cksum": "8a648e87a02fa31d9d9a3b7c76dbfee469402fbb4af3ae98b36c099d8a82bb18",
+        "features": {},
+        "yanked": false,
+        "links": null
+      })
+      .to_string(),
+    )
+    .unwrap();
```

Comment:
I'm seeing a whole lot of unwrap here. Maybe this function should just return a result.

I realize this is for testing, but it still feels sloppy. Maybe the caller can just unwrap.

Reply:
This would require require a conversion from `io::Error` to likely `anyhow::Error` which seemed unnecessary. Can you elaborate on why unwrap in this case is also sloppier than `?`?

Reply:
After some sleep I realized I was doing something silly 😅 

Updated. But still curious as to why `?` would be less sloppy in testing. I find it better to know exactly what and where something broke with virtually no digging required.

Reply:
This ended up getting rolled back to contain unwraps but I think this is better addressed in https://github.com/google/cargo-raze/pull/276